cardano-cli/src/Cardano/CLI/Slot.hs

```diff
@@ -0,0 +1,329 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Cardano.CLI.Slot
+  ( syncProgress
+
+  , Percentage(..)
+  , SyncProgress(..)
+  , SyncTolerance(..)
+  , TimeInterpreter(..)
+
+  , getSyncProgress
+  , mkTimeInterpreter
+  , neverFails
+  ) where
+
+import           Prelude
+
+import           Cardano.Slotting.Slot (EpochSize (..), SlotNo)
+import           Control.DeepSeq (NFData)
+import           Control.Exception (throwIO)
+import           Control.Monad ((>=>))
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Control.Monad.Trans.Except (ExceptT (..), runExceptT)
+import           Control.Monad.Trans.Maybe (MaybeT (..), exceptToMaybeT)
+import           Data.ByteString (ByteString)
+import           Data.Either (fromRight)
+import           Data.Hashable (Hashable)
+import           Data.Kind (Type)
+import           Data.Maybe (fromMaybe)
+import           Data.Ratio ((%))
+import           Data.Time.Clock (NominalDiffTime, UTCTime, getCurrentTime)
+import           Data.Word (Word32)
+import           GHC.Generics (Generic)
+import           GHC.Stack (HasCallStack)
+import           GHC.TypeLits (Symbol)
+import           Ouroboros.Consensus.BlockchainTime.WallClock.Types (RelativeTime (..),
+                   SystemStart (..), diffRelTime)
+import           Ouroboros.Consensus.HardFork.History.Qry (Interpreter, PastHorizonException (..),
+                   mkInterpreter, slotToWallclock)
+import           Ouroboros.Consensus.HardFork.History.Summary (neverForksSummary)
+import           Quiet (Quiet (..))
+
+import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
+import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
+
+data TimeInterpreterLog = MsgInterpreterPastHorizon
+  (Maybe String) -- ^ Reason for why the failure should be impossible
+  StartTime
+  PastHorizonException
+  deriving (Show)
+
+-- | Number of slots in a single epoch
+newtype EpochLength = EpochLength { unEpochLength :: Word32 }
+    deriving (Show, Eq, Generic)
+
+-- | Duration of a single slot.
+newtype SlotLength = SlotLength { unSlotLength :: NominalDiffTime }
```

Comment:
I'm in the midst of opening a `cardano-base` PR that will provide this definition in a new `Cardano.Slotting.Time` module.

It will also include `SystemStart`, which is the same as `StartTime`.

At the moment, both of these are exported by https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/WallClock/Types.hs but my imminent PR will move almost all of that module upstream to `cardano-base`.